### PR TITLE
Validate HS MC & SC Upgrades Test Suite

### PR DESCRIPTION
**Pull request by ryankwilliams**
# Change

This PR introduces a new test suite to run standalone (outside of osde2e) to validate hypershift management and service cluster upgrades.

Refer to the readme under the test suite folder to see further details about the test suite, how to run, etc.

Snippet from readme about the test suite:

* Before Suite
  * Validate input data
  * Identify MC/SC cluster versions (install & upgrade)
  * Get MC/SC kubeconfig's
  * Deploy ROSA Hosted Control Plane cluster to simulate customer workload
* After Suite
  * Destroy ROSA Hosted Control Plane cluster that was used for simulation
* Test Cases:
  * Perform SC upgrade
  * Verify SC post upgrade health checks
    * Runs osd-cluster-ready job
  * Verify Hosted Control Plane cluster was unaffected by the upgrade
  * Perform MC upgrade
  * Verify MC post upgrade health checks
    * Runs osd-cluster-ready job
  * Verify Hosted Control Plane cluster was unaffected by the upgrade

**Requires**

- https://github.com/openshift/osde2e-common/pull/4
- https://github.com/openshift/osde2e-common/pull/5

For [SDCICD-1028](https://issues.redhat.com//browse/SDCICD-1028)

### CODE REVIEW

File: test/mcscupgrade/mcscupgrade_test.go [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+	serviceClusterKubeConfigFile    *string
+	serviceClusterID                *string
+	serviceClusterName              *string
+	serviceClusterVersion           *semver.Version
+	serviceClusterUpgradeVersion    *semver.Version
+)
+
+var _ = BeforeSuite(func() {
+	var (
+		ctx                             = context.Background()
```
[the remark this answers is not in the document]
Comment (ryankwilliams) [reply test/mcscupgrade/mcscupgrade_test.go]: Resolved by c0653302d220f4359f0136de6f725bd3cca10b12

File: test/mcscupgrade/mcscupgrade_test.go [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+		ocmEnv                          = ocm.Integration
+		ocmToken                        = os.Getenv("OCM_TOKEN")
+		osdFleetMgmtManagementClusterID = os.Getenv("OSD_FLEET_MGMT_MANAGEMENT_CLUSTER_ID")
+		osdFleetMgmtServiceClusterID    = os.Getenv("OSD_FLEET_MGMT_SERVICE_CLUSTER_ID")
+		provisionShardID                = os.Getenv("PROVISION_SHARD_ID")
+		upgradeType                     = os.Getenv("UPGRADE_TYPE")
+	)
+
+	reportDir = pointer.String(getEnvVar("REPORT_DIR", envconf.RandomName(fmt.Sprintf("%s/mcscupgrade", os.TempDir()), 25)))
+	Expect(os.MkdirAll(*reportDir, os.ModePerm)).Error().ShouldNot(HaveOccurred(), "failed to create report directory")
```
[the remark this answers is not in the document]
Comment (ryankwilliams) [reply test/mcscupgrade/mcscupgrade_test.go]: Thanks @jbpratt! Resolved by c0653302d220f4359f0136de6f725bd3cca10b12

File: test/mcscupgrade/mcscupgrade_test.go [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+		hcpClusterID = &clusterID
+
+		kubeConfigFile, err := rosaProvider.KubeConfigFile(ctx, *hcpClusterID)
+		Expect(err).Error().ShouldNot(HaveOccurred(), "failed to get hosted control plane cluster %q kubeconfig file", hcpClusterID)
+		hcpClusterKubeConfigFile = &kubeConfigFile
+	}
+})
+
+var _ = AfterSuite(func() {
+	ctx := context.Background()
```
[the remark this answers is not in the document]
Comment (ryankwilliams) [reply test/mcscupgrade/mcscupgrade_test.go]: Resolved by c0653302d220f4359f0136de6f725bd3cca10b12

File: test/mcscupgrade/mcscupgrade_test.go [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+
+			hcpCluster.client, err = openshiftclient.NewFromKubeconfig(hcpCluster.kubeconfigFile)
+			Expect(err).ShouldNot(HaveOccurred(), "failed to construct hosted control plane cluster client")
+		}
+	})
+
+	_ = AfterAll(func(ctx context.Context) {
+		defer func() {
+			_ = osdProvider.Client.Close()
+		}()
```
Comment (jbpratt) [inline test/mcscupgrade/mcscupgrade_test.go:224]: where we create the client, you can do a `DeferCleanup(osdProvider.Client.Close)`

Comment (ryankwilliams) [reply test/mcscupgrade/mcscupgrade_test.go:233]: Thank you for this suggestion! Will make the changes.

File: test/mcscupgrade/mcscupgrade_test.go [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+}
+
+// osdClusterReadyHealthCheck verifies the osd-cluster-ready health check job is passing
+func osdClusterReadyHealthCheck(ctx context.Context, clusterClient *openshiftclient.Client, action, reportDir string) error {
+	var (
+		err        error
+		job        batchv1.Job
+		jobName    = "osd-cluster-ready"
+		namespace  = "openshift-monitoring"
+		newJobName = envconf.RandomName(fmt.Sprintf("%s-%s", jobName, action), 35)
```
Comment (jbpratt) [inline test/mcscupgrade/mcscupgrade_test.go:370]:
```suggestion
```

Can you try the `GenerateName` field in the objectmeta?

**Pull request merged**